subprojects/language-scala/src/main/java/org/gradle/api/internal/tasks/scala/ZincScalaCompilerFactory.java

```diff
@@ -62,39 +163,59 @@ static ZincScalaCompiler getCompiler(CacheRepository cacheRepository, final Iter
                 .withLockOptions(mode(FileLockManager.LockMode.OnDemand))
                 .open();
 
-        File compilerBridgeSourceJar = findFile("compiler-bridge", scalaClasspath);
+        File compilerBridgeSourceJar = findFile("compiler-bridge", hashedScalaClasspath.getClasspath());
         File bridgeJar = getBridgeJar(zincCache, scalaInstance, compilerBridgeSourceJar, sbt.util.Logger.xlog2Log(new SbtLoggerAdapter()));
-        ScalaCompiler scalaCompiler = ZincCompilerUtil.scalaCompiler(scalaInstance, bridgeJar, ClasspathOptionsUtil.auto());
+        ScalaCompiler scalaCompiler = new AnalyzingCompiler(
+            scalaInstance,
+            ZincUtil.constantBridgeProvider(scalaInstance, bridgeJar),
+            ClasspathOptionsUtil.auto(),
+            k -> scala.runtime.BoxedUnit.UNIT,
+            Option.apply(new ClassLoaderCache(COMPILER_CLASSLOADER_CACHE))
```

Comment:
Since we cannot use the `AbstractClassLoaderCache`, what are the options here?
I admit not being 100% at ease with what this exactly means.
From basic testing, it seems that replacing `new ClassLoaderCache(COMPILER_CLASSLOADER_CACHE)` with `new ClassLoaderCache(scalaInstance.loader())` works (in the sense that it allows Gradle tests to pass) though I did not check the performance impact.

An alternative is to forget about the internal Zinc class loader cache and go back to invoking `ZincCompilerUtil.scalaCompiler`.

In short, the changes done should not make `UpToDateScalaCompileIntegrationTest` fail. Note that in that test the `@Requires` on the first test method can be removed.

Reply:
Test shows that `ClassLoaderCache(scalaInstance.loader())` does not bring any noticeable gains over `ZincCompilerUtil.scalaCompiler`. The idea is pretty interesting, but it seems that compiler loads different classes than loader (bridge jar?). I also think that it is a little hackish to assume which classes AnalyzingCompiler will read.

Before we cross out AbstractClassLoaderCache... how about using reflection to load AbstractClassLoaderCache/TimeCheckingClassLoaderCache only when it is possible and otherwise use default ClassLoaderCache (which might leak some resources) or recreate it each time (which is slower by ~33%). By default Zinc 1.3.5 is used, so most of the users will benefit fully from the fix.

Reply:
Yeah, I think it makes a lot of sense to optimize for Zinc 1.3.5 since that's what is shipped by default. Is there a good reason to use older versions of Zinc?

Reply:
> Is there a good reason to use older versions of Zinc?

Not sure there, not a Scala user myself. However we try hard to not break compatibility in minors. We can certainly discuss dropping Zinc 1.2.x support for Gradle 7.0 for example.
For now, I think having the code handle both cases is best.

Let's see how the PR behaves through the full test suite.

Thanks again for being reactive and following up.